### PR TITLE
Close #48. Add prefer_beautifier_config option

### LINKED ISSUE #48: Allow beautifiers to use their own config file instead of Unibeautify config
**Important**: Read https://github.com/Unibeautify/vscode/issues/21#issuecomment-382419531 for why Unibeautify does not only use the beautifier's respective configuration files (i.e. `prettierrc` for Prettier, `eslintrc` for ESLint, etc).

---

Add option like `preferBeautifierConfig`.

# Example
When Unibeautify is iterating over the selected beautifiers and reaches beautifier Prettier and the beautifier's config is set to `preferBeautifierConfig=true` then Unibeautify should pass Prettier's config directly: https://prettier.io/docs/en/api.html#prettierresolveconfigfilepath-options

---

Do you think `prefer_beautifier_config` should be enabled or disabled by default? Read explanation at https://github.com/Unibeautify/unibeautify/issues/48#issuecomment-382422749 

[](https://api.gh-polls.com/poll/01CBCN6EQBAQ1NMH5XWXNCP847/Enabled%20by%20default/vote)
[](https://api.gh-polls.com/poll/01CBCN6EQBAQ1NMH5XWXNCP847/Disabled%20by%20default/vote)

### PR DESCRIPTION
**Pull request by Glavin001**

### CODE REVIEW

File: src/beautifier.ts [hunk 1]
```diff
@@ -358,19 +382,46 @@ export class Unibeautify {
...
+                beautifier.resolveConfig
+              ) {
+                return beautifier.resolveConfig({
+                  dependencies: dependencyManager,
+                  filePath,
+                  projectPath,
+                });
+              }
+              return Promise.resolve({});
+            })
```
Comment (szeck87) [inline src/beautifier.ts]: Can this chunk around resolving the beautifier config be moved to its own method?

**Comment by szeck87** [comment]
Please also fix lint warnings.

**Comment by Glavin001** [comment]
> Please also fix lint warnings and errors.


Out of scope for this PR, however, since I have to make edits anyway I will take a look 😃 .

```
> tslint {src,test,script}/**/*.ts
Warning: no-duplicate-case rule is deprecated. Replace your usage with the TSLint no-duplicate-switch-case rule.
Warning: no-empty-interfaces rule is deprecated. Replace your usage with the TSLint no-empty-interface rule.
Warning: no-stateless-class rule is deprecated. Replace your usage with the TSLint no-unnecessary-class rule.
Warning: no-var-self rule is deprecated. Replace your usage with the TSLint no-this-assignment rule.
Warning: valid-typeof rule is deprecated. Replace your usage with the TSLint typeof-compare rule.
```

File: src/beautifier.ts (hunk 1 above)
Comment (Glavin001) [reply src/beautifier.ts]: Agreed. A little more polish would be nice.

**Comment by szeck87** [comment]
Hmm maybe since none of the scripts in package.json have TSLint run against beautifier.ts and all the others due to this structure: `{src,test,script}/**/*.ts`

**Comment by szeck87** [comment]
Shoot, didn't mean to push to upstream.  Meant to open a PR against this branch from my own.

**Comment by Glavin001** [comment]
Awesome, thanks! No worries, I wasn't going to edit this until later tonight anyway.
Is this PR good to merge now? What's left?

File: src/beautifier.ts [hunk 2]
```diff
@@ -164,12 +164,33 @@ export interface Beautifier {
    * Runtime dependencies of the beautifier.
    */
   dependencies?: DependencyDefinition[];
+  /**
+   * Function to retrieve beautifier-specific configuration file and/or parsed value.
+   */
+  resolveConfig?(resolveConfigData: ResolveConfigData): Promise<ResolvedConfig>;
   /**
   Beautify the given code with the beautifier.
   */
```
Comment (odahcam) [inline src/beautifier.ts]: This comment is looking a little strange. 🤔

**Comment by odahcam** [comment]
What about `.editorconfig` files?

File: src/beautifier.ts (hunk 2 above)
Comment (Glavin001) [reply src/beautifier.ts]: Any particular questions you have?

**Comment by Glavin001** [comment]
> What about .editorconfig files?

EditorConfig is applicable to more than just a single beautifier.
This feature is for beautifier-specific config files, such as `.prettierrc` for Prettier and `.eslintrc` for ESLint. This is at the beautifier level.
`.editorconfig` will be coming soon and applicable at the top language level.

A possible config (work in progress):

```
JavaScript:
  import_configs: ["EditorConfig"]
```

This would be a part of https://github.com/Unibeautify/unibeautify-options-importer which still needs some design work.

The purpose of this feature ( #48 ) is to provide a way for users to use **beautifier specific** existing files, `.prettierrc` or `.eslintrc`, `.php-cs-fixer.dist`.

**Comment by szeck87** [comment]
> Is this PR good to merge now? What's left?
Ideally the `beautifyWithBeautifiers` method would be shorter.  But actually don't think moving that block of code from the review into its own method will solve it since so many params would be needed.

**Comment by Glavin001** [comment]
I'll take another look for refactoring this weekend. Lots of room for improvements. I want to make sure we are adding and allowing users to test key features. We can improve the code in the background 😃 .

**Comment by Glavin001** [comment]
Pull Requests are very welcome for any little refactoring / code improvements 😃 .

**Comment by Glavin001** [comment]
Published to v0.13.0

File: src/beautifier.ts (hunk 2 above)
Comment (odahcam) [reply src/beautifier.ts]: Shouldn't doc comments have `*` in all lines?

Comment (szeck87) [reply src/beautifier.ts]: https://api.unibeautify.com/interfaces/_beautifier_.beautifier.html#resolveconfig

Comment (odahcam) [reply src/beautifier.ts]: Actually talking about line 174's function `beautify(data: BeautifierBeautifyData)` which comment is in 171-173.

Comment (szeck87) [reply src/beautifier.ts]: It's right above resolveconfig in the docs

**Pull request merged**